### PR TITLE
160752458 Add Clojure implementation of Kalkati to GTFS conversion

### PR DESCRIPTION
**Pull request by tatut**
# Added
* New and improved Kalkati to GTFS implementation

### CODE REVIEW

File: ote/src/clj/ote/gtfs/kalkati_to_gtfs.clj [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+   routes-with-trips))
+
+(defn kalkati->gtfs
+  "Takes a parsed zipper of Kalkati.net XML and returns the equivalent
+  contents as GTFS CSV-files. Returns a sequence of GTFS files as maps
+  containing :name and :data keys."
+  [kz]
+  (let [trans-modes (trans-modes-by-id kz)
+        routes (routes-with-trips (routes kz))
+
```
Comment (rckopteri) [inline ote/src/clj/ote/gtfs/kalkati_to_gtfs.clj]: No need for these empty lines

File: database/src/main/resources/db/migration/R__GTFS_query.sql [hunk 2]
```diff
@@ -508,9 +508,9 @@ AS $$
 SELECT string_agg(rh.hash::TEXT,' ' ORDER BY h."package-id")
   FROM "gtfs-date-hash" h
   JOIN LATERAL unnest(h."route-hashes") rh
-    ON (rh."route-short-name" = route_short_name AND
-        rh."route-long-name" = route_long_name AND
-        rh."trip-headsign" = trip_headsign)
+    ON (COALESCE(rh."route-short-name",'') = COALESCE(route_short_name,'') AND
```
Comment (K1ll3rF0x) [inline database/src/main/resources/db/migration/R__GTFS_query.sql]: :+1:

File: ote/project.clj [hunk 3]
```diff
@@ -83,12 +83,12 @@
                  [org.clojure/data.zip "0.1.2"]
 
                  ;; GeoTools
-                 [org.geotools/gt-shapefile "16.1"]
-                 [org.geotools/gt-process-raster "16.1"]
-                 [org.geotools/gt-epsg-wkt "16.1"]
-                 [org.geotools/gt-geometry "16.1"]
-                 [org.geotools/gt-xml "16.1"]
-                 [org.geotools/gt-geojson "16.1"]
+                 ;[org.geotools/gt-shapefile "20.0"]
```
Comment (K1ll3rF0x) [inline ote/project.clj]: We could remove these lines completely, as we do no need these deps anymore.

File: ote/src/clj/ote/gtfs/kalkati_to_gtfs.clj [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+(defn kalkati->gtfs-zip-file
+  "Takes a parsed zipper of Kalkati.net XML and an output stream.
+  Writes the equivalent GTFS zip file to the output stream."
+  [kz output]
+  (try
+    (zip-file/write-zip (kalkati->gtfs kz) output)
+    (catch Exception e
+      (log/error e "Kalkati.net to GTFS conversion failed."))))
+
+(defn convert
```
[the remark this answers is not in the document]
Comment (K1ll3rF0x) [reply ote/src/clj/ote/gtfs/kalkati_to_gtfs.clj]: @rckopteri In clojure, you would usually just use arrow in the name if transforming some data.

File: ote/src/cljc/ote/geo.cljc [hunk 5]
```diff
@@ -174,30 +137,34 @@
...
-                (set! (.-x c) y)
-                (set! (.-y c) x))))))
+   ;; We need to define our own KKJ CRS because the one defined in geotools
+   ;; has the wrong bursa wolf parameters (the TOWGS84 line below) that resulted
+   ;; in wrong coordinates for kalkati stations (~10 - 15 meters off).
+   (def kkj
+          (org.geotools.referencing.CRS/parseWKT "PROJCS[\"KKJ / Finland Uniform Coordinate System\",
+GEOGCS[\"KKJ\",DATUM[\"Kartastokoordinaattijarjestelma_1966\",
+SPHEROID[\"International 1924\",6378388,297,AUTHORITY[\"EPSG\",\"7022\"]],
+TOWGS84[-90.7, -106.1, -119.2, 4.09, 0.218, -1.05, 1.37],
```
Comment (K1ll3rF0x) [inline ote/src/cljc/ote/geo.cljc]: We should mention the tool where to get them in case a similar need arises again: http://www.epsg-registry.org/

**Pull request merged**